gcloud/logging/handlers/transports/background_thread.py

```diff
@@ -0,0 +1,164 @@
+
+                if len(self.batch.entries) > 0:
+                    self.batch.commit()
+        finally:
+            self._entries_condition.release()
+
+        # main thread may be waiting for worker thread to finish writing its
+        # final entries. here we signal that it's done.
+        self._stop_condition.acquire()
+        self._stop_condition.notify()
+        self._stop_condition.release()
+
+    def _start(self):
+        """_start is called by this class's constructor, and is responsible
+        for starting the thread and registering the exit handlers. """
+        try:
+            self._entries_condition.acquire()
+            self._thread = threading.Thread(
+                target=self._run,
+                name="gcloud.logging.handlers.transport.Worker")
+            self._thread.setDaemon(True)
+            self._thread.start()
+        finally:
+            self._entries_condition.release()
+            atexit.register(self._stop)
+
+    def _stop(self):
+        """_stop signals the worker thread to shut down, and waits for
+        stop_timeout second for it to finish.
+
+         _stop is called by the atexit handler registered by
+         :meth:`start`. """
+        if not self.started or self.stopping:
+            return
+
+        # lock the stop condition first so that the worker
+        # thread can't notify it's finished before we wait
+        self._stop_condition.acquire()
+
+        # now notify the worker thread to shutdown
+        self._entries_condition.acquire()
+        self.stopping = True
+        self._entries_condition.notify()
+        self._entries_condition.release()
+
+        # now wait for it to signal it's finished
+        self._stop_condition.wait(self._stop_timeout)
+        self._stop_condition.release()
+        self.stopped = True
+
+    def enqueue(self, record, message):
+        """enqueue queues up a log entry to be written by the background
+        thread. """
+        try:
+            self._entries_condition.acquire()
+            if self.stopping:
+                return
+            self.batch.log_struct({"message": message,
+                                   "python_logger": record.name},
+                                  severity=record.levelname)
+            self._entries_condition.notify()
+        finally:
+            self._entries_condition.release()
+
+
+class BackgroundThreadTransport(Transport):
+    """Aysnchronous tranpsort that uses a background thread to write logging
+    entries as a batch process"""
+
+    def __init__(self, client, name):
+        super(BackgroundThreadTransport, self).__init__(client, name)
+        http = copy.deepcopy(client.connection.http)
+        http = client.connection.credentials.authorize(http)
+        self.client = Client(client.project,
+                             client.connection.credentials,
+                             http)
+        logger = self.client.logger(name)
+        self.worker = _Worker(logger)
+
```

Comment:
Traceback (most recent call last):
  File "/Users/waprin/code/gcloud-python/system_tests/logging_.py", line 162, in test_log_handler_async
    handler = CloudLoggingHandler(Config.CLIENT)
  File "/Users/waprin/code/gcloud-python/gcloud/logging/handlers/handlers.py", line 79, in __init__
    self.transport = transport(client, name)
  File "/Users/waprin/code/gcloud-python/gcloud/logging/handlers/transports/background_thread.py", line 145, in __init__
    self.client = copy.deepcopy(client)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 190, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 334, in _reconstruct
    state = deepcopy(state, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 257, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 190, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 334, in _reconstruct
    state = deepcopy(state, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 257, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 190, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 334, in _reconstruct
    state = deepcopy(state, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 257, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 190, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 334, in _reconstruct
    state = deepcopy(state, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 257, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 190, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 334, in _reconstruct
    state = deepcopy(state, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 257, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 190, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/copy.py", line 329, in _reconstruct
    y = callable(*args)
  File "/Users/waprin/.virtualenvs/docsamples10/lib/python2.7/copy_reg.py", line 93, in __newobj__
    return cls.__new__(cls, *args)
TypeError: object.__new__(_cffi_backend.CDataGCP) is not safe, use _cffi_backend.CDataGCP.__new__()